### DISCUSSION TITLE
Need Help for Command Overwrite using TypeScript

### DISCUSSION THREAD

**Discussion by Deep-Adhikary** [webdriverio/webdriverio#11758] [Q&A]
Hi Awesome Folks, 
I need help to overwrite certain commands in typescript. I have overridden `waitForClickable` as **appium does not have built in support** for it

```ts
driver.overwriteCommand('waitForClickable', async function (
    this: WebdriverIO.Element,
    {
      timeout = this.options.waitforTimeout,
      interval = this.options.waitforInterval,
      reverse = false,
      timeoutMsg = `element ("${this.selector}") still ${
        reverse ? '' : 'not '
      }clickable after ${timeout}ms`,
    }: WaitForOptions
  ) {
    return this.waitUntil(
      async () => {
        if (!this.isMobile) return reverse !== (await this.isClickable());
        const isClickable = this.isAndroid
          ? (await this.getAttribute('clickable')) === 'true'
          : (await this.getAttribute('enabled')) === 'true';
        return reverse !== isClickable;
      },
      { timeout, timeoutMsg, interval }
    );
  },
  true
);
```
#### Though the code is running perfectly, typescript is not happy with it. Can anyone tell me the desired way of command overwrite (both browser and element scope) in typescript

**Comment by christian-bromann** [comment]
> typescript is not happy with it.

How so? What is the error you are seeing?

> ```ts
>           ? (await this.getAttribute('clickable')) === 'true'
>           : (await this.getAttribute('enabled')) === 'true'
> ```

Are these common mobile attributes that are worth adding to the core command?

**Reply by Deep-Adhikary** [reply]
>How so? What is the error you are seeing?


Type script is throwing following type error. I have used `// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore` to suppress it. 

```
TS2345: Argument of type
(this: WebdriverIO.Element, { timeout, interval, reverse, timeoutMsg }: WaitForOptions) => Promise<never>
is not assignable to parameter of type
OverwriteCommandFn<"waitForClickable", "$" | "custom$" | "react$" | "$$" | "custom$$" | "react$$" | "action" | "actions" | "call" | "debug" | "deleteCookies" | "execute" | "executeAsync" | ... 24 more ... | "waitUntil", true> | OverwriteCommandFnScoped<...>
Type
(this: WebdriverIO.Element, { timeout, interval, reverse, timeoutMsg }: WaitForOptions) => Promise<never>
is not assignable to type
OverwriteCommandFn<"waitForClickable", "$" | "custom$" | "react$" | "$$" | "custom$$" | "react$$" | "action" | "actions" | "call" | "debug" | "deleteCookies" | "execute" | "executeAsync" | "getCookies" | ... 23 more ... | "waitUntil", true>
Types of parameters  __1  and  origCommand  are incompatible.
Type
(...args: any[]) => (this: Element, { timeout, interval, reverse, timeoutMsg }?: WaitForOptions | undefined) => Promise<never>
has no properties in common with type  WaitForOptions 
```

**Comment by Deep-Adhikary** [comment]
> > ```ts
> >           ? (await this.getAttribute('clickable')) === 'true'
> >           : (await this.getAttribute('enabled')) === 'true'
> > ```
> 
> Are these common mobile attributes that are worth adding to the core command?

This can be worth, but need to check if all appium drivers has the relevant properties.
`clickable` is available for **UIAutomator2**
`enabled` is available for **XCUITest**, but not sure about other drivers. Moreover, to make clickable more reliable in iOS, another attribute named `hittable` is needed which is present in **xcode** but not exposed by **XCUITest**.
